Remove LIMIT and OFFSET on relations when batch iterating

drcapulet opened the pull request:
When using `find_each`, `find_in_batches`, or `in_batches` on a relation with a specified limit, the batch relation yielded no longer propagates those values.

Other than being being useless or potentially problematic, the removal also eliminates a subquery under certain conditions when using `JOIN` with MySQL.

In the test cases I added, the queries executed go from: 

```sql
SELECT `posts`.* FROM `posts` LEFT OUTER JOIN `comments` ON `comments`.`post_id` = `posts`.`id` WHERE `comments`.`id` IS NULL ORDER BY `posts`.`id` ASC LIMIT 3
DELETE FROM `posts` WHERE `posts`.`id` IN (SELECT `id` FROM (SELECT `posts`.`id` FROM `posts` LEFT OUTER JOIN `comments` ON `comments`.`post_id` = `posts`.`id` WHERE `comments`.`id` IS NULL AND `posts`.`id` IN (3, 6, 8) LIMIT 5) AS __active_record_temp)
SELECT `posts`.* FROM `posts` LEFT OUTER JOIN `comments` ON `comments`.`post_id` = `posts`.`id` WHERE `comments`.`id` IS NULL AND `posts`.`id` > 8 ORDER BY `posts`.`id` ASC LIMIT 2
DELETE FROM `posts` WHERE `posts`.`id` IN (SELECT `id` FROM (SELECT `posts`.`id` FROM `posts` LEFT OUTER JOIN `comments` ON `comments`.`post_id` = `posts`.`id` WHERE `comments`.`id` IS NULL AND `posts`.`id` IN (9, 10) LIMIT 5) AS __active_record_temp)
```

To:

```sql
SELECT `posts`.* FROM `posts` LEFT OUTER JOIN `comments` ON `comments`.`post_id` = `posts`.`id` WHERE `comments`.`id` IS NULL ORDER BY `posts`.`id` ASC LIMIT 3
DELETE `posts` FROM `posts` LEFT OUTER JOIN `comments` ON `comments`.`post_id` = `posts`.`id` WHERE `comments`.`id` IS NULL AND `posts`.`id` IN (3, 6, 8)
SELECT `posts`.* FROM `posts` LEFT OUTER JOIN `comments` ON `comments`.`post_id` = `posts`.`id` WHERE `comments`.`id` IS NULL AND `posts`.`id` > 8 ORDER BY `posts`.`id` ASC LIMIT 2
DELETE `posts` FROM `posts` LEFT OUTER JOIN `comments` ON `comments`.`post_id` = `posts`.`id` WHERE `comments`.`id` IS NULL AND `posts`.`id` IN (9, 10).
```

And from:

```sql
SELECT `posts`.* FROM `posts` LEFT OUTER JOIN `comments` ON `comments`.`post_id` = `posts`.`id` WHERE `comments`.`id` IS NULL ORDER BY `posts`.`id` ASC LIMIT 3 OFFSET 2
UPDATE `posts` SET `posts`.`legacy_comments_count` = 0 WHERE `posts`.`id` IN (SELECT `id` FROM (SELECT `posts`.`id` FROM `posts` LEFT OUTER JOIN `comments` ON `comments`.`post_id` = `posts`.`id` WHERE `comments`.`id` IS NULL AND `posts`.`id` IN (8, 9, 10) LIMIT 18446744073709551615 OFFSET 2) AS __active_record_temp)
SELECT `posts`.* FROM `posts` LEFT OUTER JOIN `comments` ON `comments`.`post_id` = `posts`.`id` WHERE `comments`.`id` IS NULL AND `posts`.`id` > 10 ORDER BY `posts`.`id` ASC LIMIT 3 OFFSET 2
```

To: 

```sql
SELECT `posts`.* FROM `posts` LEFT OUTER JOIN `comments` ON `comments`.`post_id` = `posts`.`id` WHERE `comments`.`id` IS NULL ORDER BY `posts`.`id` ASC LIMIT 3 OFFSET 2
UPDATE `posts` LEFT OUTER JOIN `comments` ON `comments`.`post_id` = `posts`.`id` SET `posts`.`legacy_comments_count` = 0 WHERE `comments`.`id` IS NULL AND `posts`.`id` IN (8, 9, 10)
SELECT `posts`.* FROM `posts` LEFT OUTER JOIN `comments` ON `comments`.`post_id` = `posts`.`id` WHERE `comments`.`id` IS NULL AND `posts`.`id` > 10 ORDER BY `posts`.`id` ASC LIMIT 3 OFFSET 2.
```

While both can cause improper query execution, the test cases shows the usage of `OFFSET` is much easier to trigger this issue compared to `LIMIT`.

I believe the subquery behavior with Postgres and SQLite is as expected, but I'm not as familiar with their internals.

jeromedalbert commented:
@drcapulet Can you confirm this is still an issue? If so, can you maybe provide a reproduction [script](https://github.com/rails/rails/tree/main/guides/bug_report_templates)?

drcapulet committed:
Remove LIMIT and OFFSET on relations when batch iterating

When using find_each, find_in_batches, or in_batches on a relation with a specified limit, the batch relation yielded no longer propagates those values.

Other than being being useless or potentially problematic, the removal also eliminates a subquery under certain conditions when using JOIN with MySQL.

fatkodima commented:
> Unfortunately with https://github.com/rails/rails/pull/52385 I'm not sure the best way to land this

I don't get how it that PR relevant?

CI is red. Also update tests to test all adapters, because they all will have the same behavior. And avoid using vertical alignment (noticed in tests).
